diag_manager/fms_diag_object_container.F90

```diff
@@ -0,0 +1,201 @@
+   function remove_diag_object (this, key, iiter ) result (riter)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer , intent (in) :: key
+      class(diag_obj_iterator), intent (in), optional :: iiter
+      class(diag_obj_iterator), allocatable:: riter
+      class(literator), allocatable :: temp_liter
+      integer :: status
+      !!
+      !Find the object in the container.
+      riter = this%find ( key , iiter)
+      temp_liter = this%the_ll%remove( riter%liter%current )
+      riter = diag_obj_iterator(temp_liter)
+   end function
+
+   !> return the number of objects held in the container
+   function get_num_objects (this )result (sz)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer :: sz
+      sz = this%the_ll%size()
+   end function
+
+
+   !> return a pointer to an iterator for the objects in the continer.
+   function get_iterator (this) result (oliter)
+      class (fms_diag_object_container), intent (in) :: this
+      class(diag_obj_iterator) , allocatable :: oliter
+      oliter = diag_obj_iterator( this%the_ll%get_literator() )
+   end function
+
+
+   function diag_obj_iterator_constructor( iliter ) result (diag_itr)
+    class (literator), allocatable :: iliter
+    class (diag_obj_iterator), allocatable :: diag_itr
+    allocate(diag_itr)
+    diag_itr%liter = iliter;
+ end function diag_obj_iterator_constructor
+
+
+   function diag_object_container_constructor () result (doc)
+      type(fms_diag_object_container), allocatable :: doc
+      allocate(doc)
+      doc%the_ll = linked_list()
+      print * , "In DOC constructor"
+   end function diag_object_container_constructor
+
+
+
+   !> Return true if there is more data this iterator can access
+   function literator_has_data( this ) result( r )
+      class(diag_obj_iterator), intent(in) :: this
+      logical :: r
+      r = this%liter%has_data()
+   end function literator_has_data
+
+   !> Move the iterator to the next data
+   function literator_next( this ) result( status )
+      class(diag_obj_iterator), intent(in out ) :: this
+      integer :: status
+      status = 0
+      status = this%liter%next()
+   end function literator_next
+
+   !> Get the current data the iterator is poiting to
+   function  literator_data( this ) result( rdo )
+      class(diag_obj_iterator), intent(in) :: this
+      class(fms_diag_object),  pointer :: rdo !!the resultant diagnostic object
+      class(*),  pointer :: gp !! generic typed object in the container
+
+      gp => this%liter%get()
+      select type(gp)
+       type is (fms_diag_object)  !! "type is", not the (polymorphic) "class is"
+         rdo => gp
+       class default
+         print * , "error iter data type error"
+      end select
+   end function literator_data
+
+   subroutine destructor(this)
+      type(fms_diag_object_container) :: this  !Note for destructors its needs to be type and not class!
```

Comment:
Subroutine needs doxygen description.
Variables need to be documented in doxygen style

Reply:
done